### PR TITLE
feat/bngp-3479-payment-fee

### PR DESCRIPTION
**Pull request by sfitz42**

### CODE REVIEW

File: packages/webapp/src/routes/developer/upload-metric-file.js [hunk 1]
```diff
@@ -30,6 +30,8 @@ async function processSuccessfulUpload (result, request, h) {
   request.yar.set(constants.redisKeys.DEVELOPER_METRIC_DATA, result[0].metricData)
   request.yar.set(constants.redisKeys.DEVELOPER_METRIC_FILE_NAME, result.filename)
   logger.log(`${new Date().toUTCString()} Received metric data for ${result[0].location.substring(result[0].location.lastIndexOf('/') + 1)}`)
+  console.log(request.yar.get(constants.redisKeys.BIODIVERSITY_NET_GAIN_NUMBER))
+  console.log(result[0].metricData)
```
Comment (KiranVarma02) [inline packages/webapp/src/routes/developer/upload-metric-file.js:34]: remove?

File: packages/webapp/src/routes/developer/tasklist.js [hunk 2]
```diff
@@ -19,6 +19,7 @@ const handlers = {
           })
         }
       })
+      console.log(dataContent)
```
Comment (KiranVarma02) [inline packages/webapp/src/routes/developer/tasklist.js:22]: remove?

File: packages/webapp/src/utils/application-validation.js [hunk 3]
```diff
@@ -55,7 +55,13 @@ const applicationValidation = Joi.object({
       then: Joi.valid('true'),
       otherwise: Joi.valid('true', 'false')
     }).default('false'),
-    metricData: Joi.object().allow(null)
+    metricData: Joi.object().allow(null),
+    payment: Joi.object({
+      caseType: Joi.string().required(),
+      fee: Joi.number().required(),
+      reference: Joi.string().allow('', null).optional(),
```
[the remark this answers is not in the document]
Comment (sfitz42) [reply packages/webapp/src/utils/application-validation.js:62]: At the point of adding the payment object into the application the reference has not been generated. Therefore, I've had to add it into the schema as optional for the time being.

```js
const { value, error } = applicationValidation.validate(application(request.yar))
if (error) {
  throw new Error(error)
}
const result = await postJson(`${constants.AZURE_FUNCTION_APP_URL}/processapplication`, value)
request.yar.set(constants.redisKeys.APPLICATION_REFERENCE, result.gainSiteReference)
return h.redirect(constants.routes.REGISTRATION_SUBMITTED)
```

Not sure if you have any other suggestions? I suppose we might not actually need the reference in the payment object as the application reference for both journeys is stored in session. However, I think we were planning on having the reference eventually for GOV Pay.

File: packages/webapp/src/routes/developer/upload-metric-file.js (hunk 1 above)
Comment (sfitz42) [reply packages/webapp/src/routes/developer/upload-metric-file.js]: Done

File: packages/webapp/src/utils/application-validation.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (KiranVarma02) [reply packages/webapp/src/utils/application-validation.js:62]: please get opinion from someone who knows more on this topic, I am not sure!

[the remark this answers is not in the document]
Comment (steve-dickinson) [reply packages/webapp/src/utils/application-validation.js:62]: @KiranVarma02 With there been two diferent payment methods BACs and Gov pay, we need the payload to contain the properties required to capture the pay information. The reference is an optional property to capture the unquie identifier returned from gov pay to aid in reconciliation and querying of the payment. BACs doesn't require a reference.

[the remark this answers is not in the document]
Comment (KiranVarma02) [reply packages/webapp/src/utils/application-validation.js:62]: just for my understanding - in BACs payments without reference, how will we identify which payment is for what? 

sorry I am not sure if my questions is valid but would like to understand

**Review by steve-dickinson — APPROVED** [review]
LGTM

**Comment by shaundefra** [comment]
tested by Leena

**Pull request merged**